src/lib/OpenEXR/ImfContextInit.cpp

```diff
@@ -0,0 +1,207 @@
+
+#include <cinttypes>
+#include <climits>
+#include <cstring>
+#include <mutex>
+
+OPENEXR_IMF_INTERNAL_NAMESPACE_SOURCE_ENTER
+
+struct istream_holder
+{
+    istream_holder (IStream* s) : _stream (s)
+    {
+        if (_stream) _cur_offset = _stream->tellg ();
+    }
+
+#if ILMTHREAD_THREADING_ENABLED
+    std::mutex _mx;
+#endif
+    uint64_t _cur_offset = 0;
+    IStream* _stream;
+};
+
+static int64_t
+istream_read (
+    exr_const_context_t         ctxt,
+    void*                       userdata,
+    void*                       buffer,
+    uint64_t                    sz,
+    uint64_t                    offset,
+    exr_stream_error_func_ptr_t error_cb)
+{
+    istream_holder* ih = static_cast<istream_holder*> (userdata);
+
+    if (sz > INT_MAX)
+    {
+        error_cb (
+            ctxt,
+            EXR_ERR_READ_IO,
+            "Stream interface request to read block too large");
+        return -1;
+    }
+
+#if ILMTHREAD_THREADING_ENABLED
+    std::lock_guard<std::mutex> lk{ih->_mx};
+#endif
+
+    if (offset != ih->_cur_offset)
+    {
+        ih->_stream->seekg (offset);
+        ih->_cur_offset = ih->_stream->tellg ();
+        if (offset != ih->_cur_offset)
+        {
+            error_cb (
+                ctxt,
+                EXR_ERR_READ_IO,
+                "Unable to seek to desired offset %" PRIu64,
+                offset);
+            return -1;
+        }
+    }
+
+    int64_t nread = ih->_cur_offset;
+    try
+    {
+        ih->_stream->read (
+            static_cast<char*> (buffer), static_cast<int> (sz));
+    }
+    catch (...)
+    {
+        // bah, there could be two reasons for this, one is a
+        // legitimate error, the other is a read past the end of file
+        // (i.e. the core library tries to read a 4k block when
+        // parsing the header), let's let the core deal with that and
+        // clear errors
+        ih->_stream->clear ();
+        //error_cb (
+        //    ctxt,
+        //    EXR_ERR_READ_IO,
+        //    "Unable to read requested bytes: %" PRIu64,
+        //    sz);
+    }
+    ih->_cur_offset = ih->_stream->tellg ();
+    nread           = ih->_cur_offset - nread;
+    return nread;
+}
+
+static void
+istream_destroy (exr_const_context_t ctxt, void* userdata, int failed)
+{
+    istream_holder* ih = static_cast<istream_holder*> (userdata);
+    delete ih;
+}
+
+ContextInitializer&
+ContextInitializer::setInputStream (IStream* istr)
+{
+    _initializer.user_data  = new istream_holder{istr};
+    _initializer.read_fn    = istream_read;
+    _initializer.size_fn    = nullptr;
+    _initializer.write_fn   = nullptr;
+    _initializer.destroy_fn = istream_destroy;
+    _ctxt_type              = ContextFileType::READ;
+    _prov_stream            = istr;
+    return *this;
+}
+
+struct ostream_holder
+{
+    ostream_holder (OStream* s) : _stream (s)
+    {
+        if (_stream) _cur_offset = _stream->tellp ();
+    }
+
+#if ILMTHREAD_THREADING_ENABLED
```

Comment:
since we are moving on to effectively EXR 4, is it worth preserving the THREADING_ENABLED option?  We introduced it over 20 years ago because I wanted to run EXR on game consoles that didn't have threading capabilities. But I don't think there are any reasonable targets left that don't support threading, especially since we have moved to c++17 as a baseline requirement.

Reply:
/shrug for embedded builds with musl c or something? I know pthread isn't a separate library in most standard libc now...

Reply:
I mean I'd point those applications to Core anyway